src/sql/explain.rs

```diff
@@ -0,0 +1,482 @@
+use expr::{Id, IdHumanizer};
+use repr::{RelationType, ScalarType};
+use std::collections::{BTreeMap, HashMap};
+
+#[derive(Debug)]
+pub struct Explanation<'a> {
+    /// One ExplanationNode for each RelationExpr in the plan, in left-to-right post-order
+    pub nodes: Vec<ExplanationNode<'a>>,
+}
+
+#[derive(Debug)]
+pub struct ExplanationNode<'a> {
+    /// The expr being explained
+    pub expr: &'a RelationExpr,
+    /// The parent of expr
+    pub parent_expr: Option<&'a RelationExpr>,
+    /// A pretty-printed representation of the expr
+    pub pretty: String,
+    /// A list of annotations containing extra information about this expr
+    pub annotations: Vec<String>,
+    /// Nodes are grouped into chains of linear operations for easy printing
+    pub chain: usize,
+    /// Nodes with subqueries have a nested explanation for the subquery
+    pub subqueries: Vec<Explanation<'a>>,
+}
+
+impl<'a> std::fmt::Display for Explanation<'a> {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+        let mut prev_chain = usize::max_value();
+        for node in &self.nodes {
+            if node.chain != prev_chain {
+                if node.chain != 0 {
+                    writeln!(f)?;
+                }
+                writeln!(f, "{} =", node.chain)?;
+            }
+            prev_chain = node.chain;
+
+            // explain output shows up in SLT where the linter will not allow trailing whitespace, so need to trim stuff
+            writeln!(f, "| {}", node.pretty.trim())?;
+            for annotation in &node.annotations {
+                writeln!(f, "| | {}", annotation.trim())?;
+            }
+            for subquery in &node.subqueries {
+                for line in subquery.to_string().split('\n') {
+                    if line.is_empty() {
+                        writeln!(f, "| |")?;
+                    } else {
+                        writeln!(f, "| | {}", line)?;
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl RelationExpr {
+    /// Create an Explanation, to which annotations can be added before printing
+    pub fn explain(&self, id_humanizer: &impl IdHumanizer) -> Explanation {
+        self.explain_internal(id_humanizer, &mut 0)
+    }
+
+    fn explain_internal(
+        &self,
+        id_humanizer: &impl IdHumanizer,
+        next_chain: &mut usize,
+    ) -> Explanation {
+        use RelationExpr::*;
+
+        // get nodes in post-order
+        let mut nodes = vec![];
+        let mut stack: Vec<(Option<&RelationExpr>, &RelationExpr)> = vec![(None, self)];
+        while let Some((parent_expr, expr)) = stack.pop() {
+            nodes.push(ExplanationNode {
+                expr,
+                parent_expr,
+                pretty: String::new(),
+                annotations: vec![],
+                chain: 0,           // will fix this up later
+                subqueries: vec![], // will fix this up later
```

Comment:
as in "TODO(jamii)" or as in "this is fixed up in the code later"?

Reply:
The latter. Will make that clearer.